lib/lightning/work_orders.ex

```diff
@@ -491,6 +468,48 @@ defmodule Lightning.WorkOrders do
     {:ok, 0, 0}
   end
 
+  def get_last_runs_steps_with_dataclips(workorders, jobs) do
+    job_ids = Enum.map(jobs, & &1.id)
+
+    workorders
+    |> Enum.map(& &1.id)
+    |> last_run_steps_query(job_ids, non_wiped_dataclip?: true)
+    |> Repo.all()
+  end
+
+  defp last_run_steps_query(workorder_ids, job_ids, opts \\ []) do
+    last_runs_query = from(r in Run,
+        where: r.work_order_id in ^workorder_ids,
+        group_by: [r.work_order_id],
+        select: %{
+          work_order_id: r.work_order_id,
+          last_inserted_at: max(r.inserted_at)
+        }
+      )
+
+    from(rs in RunStep,
+        join: r in assoc(rs, :run),
+        join: s in assoc(rs, :step),
+        join: wo in assoc(r, :work_order),
+        join: lr in subquery(last_runs_query),
+        on:
+          r.work_order_id == lr.work_order_id and
+            r.inserted_at == lr.last_inserted_at,
+        where: s.job_id in ^job_ids,
+        order_by: [asc: wo.inserted_at],
+        select: %{rs | step: s}
```

Comment:
Instead of merging the RunStep struct, I think this should be a `preload: [step: s]` right?

Reply:
Preloads are tricky, there are times when they produce an additional/detached query like in this case.

Reply:
I have confirmed with debug log level that it's really sending one more query down the wire.